invenio_rdm_records/services/services.py

```diff
@@ -131,3 +148,229 @@ def oai_result_item(self, identity, oai_record_source):
             record,
             links_tpl=self.links_item_tpl,
         )
+
+    @unit_of_work()
+    def create_user_access_request(
+        self, identity, id_, message, expand=False, uow=None
+    ):
+        """Create a user access request for the given record."""
+        # TODO hide behind "authenticated_user" permission wall
+        # TODO deny if the identity already has access?
+        # TODO detect duplicate requests
+        record = self.record_cls.pid.resolve(id_)
+        self.require_permission(identity, "read", record=record)  # TODO
+
+        # TODO check that the user cannot actually access files
+        #      -> probably better with a new separate permission policy entry
+        denied = False
+        try:
+            self.require_permission(identity, "read_files", record=record)
+        except PermissionDeniedError:
+            denied = True
+
+        if not denied:
+            raise PermissionDeniedError()
+
+        # TODO try to find a duplicate request
+        req_cls = current_requests_service.record_cls
+        model_cls = req_cls.model_cls
+        requests = [
+            req_cls(rm.data, model=rm)
+            for rm in model_cls.query.filter(
+                model_cls.json["created_by"] == {"user": str(identity.id)},
+                model_cls.json["topic"] == {"record": id_},
+            )
+            if rm.data and rm.data["type"] == UserAccessRequest.type_id
+        ]
+
+        if any((req.is_open for req in requests)):
+            # TODO DuplicateRequestError
+            raise PermissionDeniedError()
+
+        record = self.record_cls.pid.resolve(id_)
+        data = {
+            "payload": {
+                "permission": record.parent.access.settings.default_access_permission,
+                "message": message,
+            }
+        }
+
+        # TODO might also be community managers
+        receiver = None
+        record_owner = record.parent.access.owner.resolve()
+        if record_owner:
+            receiver = record_owner
+
+        if receiver is None:
+            # TODO as fallback maybe admins?
+            pass
+
+        request = current_requests_service.create(
+            identity,
+            data,
+            UserAccessRequest,
+            receiver,
+            topic=record,
+            expires_at=None,  # TODO expiration date
+            expand=expand,
+            uow=uow,
+        )
+
+        # immediately submit the request, unless it has errors
+        # TODO is this correct?
```

Comment:
Yes, that's the workflow for now.
I'm just wondering if it's "correct" do simply return the created (but not yet submitted) request in case the creation step already failed.
Probably this should be wrapped in an exception instead?